plugins/eni/engine/dhclient.go

```diff
@@ -0,0 +1,283 @@
+	// leases file for a network interface. It's composed of:
+	// i. Path prefix
+	// ii. Device Name
+	// iii. Internet Protocol Revision ('4' or '6')
+	dhclientPIDFileFormat = "%s/ns-%s-dhclient%d.pid"
+
+	ipRev4 = 4
+	ipRev6 = 6
+
+	// processFinishedErrorMessage reflects the corresponding error message
+	// emitted by the os package when it finds that a process has finished
+	// its execution. Unfortunately, this is local variable in the os
+	// package and is not exposed anywhere. Hence, we declare this string to
+	// match the error message returned by the os package
+	processFinishedErrorMessage = "os: process already finished"
+)
+
+// DHClient interface provides methods to perform various dhclient actions
+type DHClient interface {
+	// IsExecutableInPath returns true if the dhclient executable is found
+	// in the path
+	IsExecutableInPath() bool
+	// Start starts the dhclient process for a given device name. The IP
+	// revision can be either '4' or '6'
+	Start(deviceName string, ipRev int) error
+	// Stop stops the dhclient process for a given device name until max
+	// stop wait duration. The IP revision can be either '4' or '6'.
+	Stop(deviceName string, ipRev int, checkStateInterval time.Duration, maxWait time.Duration) error
+}
+
+type dhclient struct {
+	os             oswrapper.OS
+	exec           execwrapper.Exec
+	ioutil         ioutilwrapper.IOUtil
+	executable     string
+	leasesFilePath string
+	pidFilePath    string
+}
+
+// NewDHClient creates a new DHClient object
+func NewDHClient() *dhclient {
+	return newDHClient(
+		oswrapper.NewOS(),
+		execwrapper.NewExec(),
+		ioutilwrapper.NewIOUtil())
+}
+
+func newDHClient(os oswrapper.OS, exec execwrapper.Exec, ioutil ioutilwrapper.IOUtil) *dhclient {
+	client := &dhclient{
+		os:     os,
+		exec:   exec,
+		ioutil: ioutil,
+	}
+	client.loadEnvConfig()
+	return client
+}
+
+func (client *dhclient) loadEnvConfig() {
+	client.executable = client.getEnvOrDefault(
+		dhclientExecutableNameEnvConfig, dhclientExecutableNameDefault)
+	client.leasesFilePath = client.getEnvOrDefault(
+		dhclientLeasesFilePathEnvConfig, dhclientLeasesFilePathDefault)
+	client.pidFilePath = client.getEnvOrDefault(
+		dhclientPIDFilePathEnvConfig, dhclientPIDFilePathDefault)
+}
+
+func (client *dhclient) getEnvOrDefault(key string, defaultValue string) string {
+	envVal := client.os.Getenv(key)
+	if envVal == "" {
+		return defaultValue
+	}
+	return envVal
+}
+
+func (client *dhclient) IsExecutableInPath() bool {
+	dhclientPath, err := client.exec.LookPath(client.executable)
+	if err != nil {
+		log.Warnf("Error searching dhclient in PATH: %v", err)
+		return false
+	}
+
+	log.Debugf("dhclient found in: %s", dhclientPath)
+	return true
+}
+
+func (client *dhclient) Start(deviceName string, ipRev int) error {
+	args := client.constructDHClientArgs(deviceName, ipRev)
+	cmd := client.exec.Command(client.executable, args...)
+	out, err := cmd.CombinedOutput()
```

Comment:
@nmeyerhans Thanks for pointing this out. You're right that we do not have enforce any timeouts here beyond what's set as the default. I tried various options mentioned the man page for `dhclient`, but was unable to make the process say in foreground and exit after a timeout interval. I think that enforcing a timeout outside of the `dhclient` command via the invocation code might be an acceptable solution here where we mark it as an error if the timeout is breached. I am in favor of doing this in another PR though. Let me know if that sounds ok to you.